File: Objective-C/AdaptiveCell/Controller/ViewController.m
//
//  ViewController.m
//  AdaptiveCell
//
//  Created by smilingmiao on 2019/3/18.
//  Copyright © 2019 smilingmiao. All rights reserved.
//

#import "ViewController.h"
#import "AdaptiveTableViewCell.h"
#import "PureCodeTableViewCell.h"
#import "FeedHolder.h"
#import "CommonData.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) FeedHolder *feedHolder;
@property (nonatomic, assign) BOOL isUseXib;

@property (nonatomic, weak) UITextField *bookTextField;
@property (nonatomic, weak) UITextField *summaryTextField;
@property (nonatomic, weak) UIAlertAction *confirmAction;

@end

@implementation ViewController

static NSString * const cellID = @"adaptiveCell";

#pragma mark - life circle
- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.title = @"Books";
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(addItem)];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(handleFeedsChange:)
                                                 name:kFeedsChangedNotificationName
                                               object:nil];
    
    [self.view addSubview:self.tableView];

    /// 注册 xib 还是纯代码，其中纯代码又分用 Masonry 和用 iOS 自带 AutoLayout，这个需要在 `PureCodeTableViewCell` 中设置。
    [self registerWithXib:YES];
    
    self.feedHolder = [[FeedHolder alloc] init];
    [self.feedHolder buildData];
}

- (void)viewWillLayoutSubviews
{
    [super viewWillLayoutSubviews];
    
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addConstraint:[self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor]];
    [self.view addConstraint:[self.tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor]];
    [self.view addConstraint:[self.tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor]];
    [self.view addConstraint:[self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor constant:-35]];
}

#pragma mark - private method
- (void)registerWithXib:(BOOL)isXib
{
    if (isXib) {
        UINib *nib = [UINib nibWithNibName:@"AdaptiveTableViewCell" bundle:nil];
        [self.tableView registerNib:nib forCellReuseIdentifier:cellID];
    } else {
        [self.tableView registerClass:[PureCodeTableViewCell class]
               forCellReuseIdentifier:cellID];
    }
    _isUseXib = isXib;
}

- (void)addItem
{
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"增加条目" message:@"填写书名和简介" preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"请填写书名";
        [textField addTarget:self action:@selector(inputText:) forControlEvents:UIControlEventEditingChanged];
        self.bookTextField = textField;
    }];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"请填写简介";
        [textField addTarget:self action:@selector(inputText:) forControlEvents:UIControlEventEditingChanged];
        self.summaryTextField = textField;
    }];
    [alert addAction:[UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
        [self dismissViewControllerAnimated:YES completion:nil];
    }]];
    
    self.confirmAction = [UIAlertAction actionWithTitle:@"提交" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        NSDictionary *dict = @{@"title": self.bookTextField.text,
                               @"content": self.summaryTextField.text};
        Feed *feed = [[Feed alloc] initWithDictionary:dict];
        [self.feedHolder addFeed:feed atIndex:0];
    }];
    [self.confirmAction setValue:[UIColor blueColor] forKey:@"titleTextColor"];
    self.confirmAction.enabled = NO;
    [alert addAction:self.confirmAction];
    
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)inputText:(UITextField *)textField
{
    if (self.bookTextField.text.length > 0 && self.summaryTextField.text.length > 0) {
        self.confirmAction.enabled = YES;
    } else {
        self.confirmAction.enabled = NO;
    }
}

- (void)deleteAtIndex:(NSUInteger)index
{
    if (self.feedHolder.count > index) {
        [self.feedHolder removeAtIndex:index];
    }
}

- (void)handleFeedsChange:(NSNotification *)notification
{
    Behavior *behavior = (Behavior *)notification.userInfo[kFeedHolderDidChangeBehaviorKey];
    [self syncTableView:behavior];
}

- (void)syncTableView:(Behavior *)behavior
{
    switch (behavior.behaviorType) {
        case InsertRow: {
            NSIndexPath *indexPath = [NSIndexPath indexPathForRow:behavior.index inSection:0];
            [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationTop];
            break;
        }
        case deleteRow: {
            NSIndexPath *indexPath = [NSIndexPath indexPathForRow:behavior.index inSection:0];
            [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationBottom];
            break;
        }
        case reloadRows: {
            [self.tableView reloadData];
            break;
        }
        default: {
            break;
        }
    }
}

#pragma mark - table view data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return (NSInteger)self.feedHolder.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = nil;
    if (_isUseXib) {
        cell = [tableView dequeueReusableCellWithIdentifier:cellID forIndexPath:indexPath];
        [(AdaptiveTableViewCell *)cell setFeed:[self.feedHolder feed:indexPath.row]];
    } else {
        cell = [tableView dequeueReusableCellWithIdentifier:cellID forIndexPath:indexPath];
        if (cell == nil) {
            cell = [[PureCodeTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellID];
        }
        [(PureCodeTableViewCell *)cell bindData:[self.feedHolder feed:indexPath.row]];
    }
    
    return cell;
}

#pragma mark - table view delegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewAutomaticDimension;
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UIContextualAction *deleteAction = [UIContextualAction contextualActionWithStyle:UIContextualActionStyleNormal title:@"删除" handler:^(UIContextualAction * _Nonnull action, __kindof UIView * _Nonnull sourceView, void (^ _Nonnull completionHandler)(BOOL)) {
        [self deleteAtIndex:indexPath.row];
        completionHandler(YES);
    }];
    deleteAction.backgroundColor = [UIColor redColor];
    
    return [UISwipeActionsConfiguration configurationWithActions:@[deleteAction]];
}

#pragma mark - getter & setter
- (UITableView *)tableView
{
    if (_tableView == nil) {
        _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
        _tableView.contentInset = UIEdgeInsetsMake(0, 0, 0, 0);
        _tableView.delegate = (id)self;
        _tableView.dataSource = (id)self;
        _tableView.estimatedRowHeight = 100;
        _tableView.sectionHeaderHeight = 0;
        _tableView.sectionFooterHeight = 0;
    }
    return _tableView;
}

@end

File: Swift/AdaptiveCell/ViewController/ViewController.swift
//
//  ViewController.swift
//  AdaptiveCell
//
//  Created by smilingmiao on 2019/3/30.
//  Copyright © 2019 smilingmiao. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
  
  private var tableView: UITableView!
  private var store: FeedStore!
  
  private weak var bookTextField: UITextField?
  private weak var summaryTextField: UITextField?
  private weak var confirmAction: UIAlertAction?
  
  override func viewDidLoad() {
    super.viewDidLoad()
    navigationItem.title = "Books"
    navigationItem.rightBarButtonItem = .init(barButtonSystemItem: .add, target: self, action: #selector(addBook))
    
    NotificationCenter.default.addObserver(self, selector: #selector(feedItemDidChange), name: .feedStoreDidChangeNotification, object: nil)
    
    setupTableView()
    store = FeedStore()
  }
  
  override func viewWillLayoutSubviews() {
    super.viewWillLayoutSubviews()
    
    tableView.translatesAutoresizingMaskIntoConstraints = false
    view.addConstraint(tableView.topAnchor.constraint(equalTo: view.topAnchor, constant: 0))
    view.addConstraint(tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 10))
    view.addConstraint(tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -10))
    let margin = UIScreen.main.bounds.size.height >= 812 ? CGFloat(35): CGFloat(0)
    view.addConstraint(tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -margin))
  }
  
  private func setupTableView() {
    tableView = UITableView(frame: .zero, style: .plain)
    tableView.contentInset = .init(top: 0, left: 0, bottom: 0, right: 0)
    tableView.dataSource = self
    tableView.delegate = self
    tableView.estimatedRowHeight = 100
    tableView.sectionHeaderHeight = 0
    tableView.sectionFooterHeight = 0
    
    let nibName = UINib(nibName: "AdaptiveTableViewCell", bundle: nil)
    tableView.register(nibName, forCellReuseIdentifier: "FeedCell")
    
    view.addSubview(tableView)
  }
  
  @objc func addBook(sender: Any) {
    
    let alert = UIAlertController(title: "增加条目", message: "填写书名和简介", preferredStyle: .alert)
    
    alert.addTextField { (textField) in
      textField.placeholder = "请填写书名"
      textField.addTarget(self, action: #selector(self.inputText), for: .editingChanged)
      self.bookTextField = textField
    }
    alert.addTextField { (textField) in
      textField.placeholder = "请填写简介"
      textField.addTarget(self, action: #selector(self.inputText), for: .editingChanged)
      self.summaryTextField = textField
    }
    alert.addAction(UIAlertAction(title: "取消", style: .cancel, handler: { (_) in
      self.dismiss(animated: true, completion: nil)
    }))
    
    let confirm = UIAlertAction(title: "提交", style: .destructive) { (_) in
      let title = self.bookTextField?.text ?? ""
      let content = self.summaryTextField?.text ?? ""
      let feed = FeedItem(title: title, content: content)
      self.store.add(item: feed, at: 0)
    }
    alert.addAction(confirm)
    confirm.setValue(UIColor.blue, forKey: "titleTextColor")
    confirm.isEnabled = false
    confirmAction = confirm
    
    self.present(alert, animated: true, completion: nil)
  }
  
  @objc func inputText(_ textField: UITextField) {
    let title = bookTextField?.text ?? ""
    let content = summaryTextField?.text ?? ""
    if title.count > 0 && content.count > 0 {
      confirmAction?.isEnabled = true
    } else {
      confirmAction?.isEnabled = false
    }
  }
  
  @objc func feedItemDidChange(_ notification: Notification) {
    let behavior = notification.getUserInfo(for: .feedStoreDidChangedChangeBehaviorKey)
    switch behavior {
    case .add(let index):
      let indexPathes = index.map { IndexPath(row: $0, section: 0) }
      self.tableView.insertRows(at: indexPathes, with: .automatic)
      
    case .remove(let index):
      let indexPathes = index.map { IndexPath(row: $0, section: 0) }
      tableView.deleteRows(at: indexPathes, with: .automatic)
    case .reload:
      tableView.reloadData()
    }
  }
  
  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return store.count
  }
  
  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    if let cell = tableView.dequeueReusableCell(withIdentifier: "FeedCell", for: indexPath) as? AdaptiveTableViewCell {
      if indexPath.row >= 0 && indexPath.row < store.count {
        cell.bindData(current: store.item(at: indexPath.row))
      }
      return cell
    }
    return UITableViewCell()
  }
  
  func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
    return UITableView.automaticDimension
  }
  
  func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
    let contextualAction = UIContextualAction(style: .normal, title: "删除") { (_, _, done) in
      self.store.remove(at: indexPath.row)
      done(true)
    }
    contextualAction.backgroundColor = .orange
    let configuration = UISwipeActionsConfiguration(actions: [contextualAction])
    configuration.performsFirstActionWithFullSwipe = false
    
    return configuration
  }
  
  
}
